File: src/main/java/view25d/prims/SetObserverAngles.java
package view25d.prims;

import org.nlogo.api.Argument;
import org.nlogo.api.Context;
import org.nlogo.api.DefaultCommand;
import org.nlogo.api.ExtensionException;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;

import view25d.View25DExtension;
import view25d.view.MouseableGLWindow;
import view25d.view.VarviewWindow;

public class SetObserverAngles extends DefaultCommand {

	@Override
    public Syntax getSyntax() {
       int[] argTypes = {Syntax.StringType(), Syntax.NumberType(), Syntax.NumberType()};
       return  Syntax.commandSyntax(argTypes);
    }
	
	@Override
	public String getAgentClassString() {
		return "O";
	}
	
	@Override
	public void perform(Argument[] args, Context context)
			throws ExtensionException, LogoException {
		
		String whichView = args[0].getString();
		double newHeading = args[1].getDoubleValue();
		double newPitch = args[2].getDoubleValue();
		
		VarviewWindow vvwp = View25DExtension.patchWindowMap.get(whichView);
		
 		VarviewWindow vvwt = View25DExtension.turtleWindowMap.get(whichView);
 		
		if (vvwp != null) {
			MouseableGLWindow win = vvwp.getGLWindow();
			double [] headingAndPitch = win.getObserverPerspectiveAngles();
			
			double oldHeading = headingAndPitch[0];
			double delHeading = newHeading - oldHeading;
			
			double oldPitch = headingAndPitch[1];
			double delPitch = -1 * (newPitch - oldPitch);

			win.updateObserverPerspectiveAnglesWithDeltas(delHeading, delPitch);
		}
		
		if (vvwt != null) {
			MouseableGLWindow win = vvwt.getGLWindow();
			double [] headingAndPitch = win.getObserverPerspectiveAngles();
			
			double oldHeading = headingAndPitch[0];
			double delHeading = newHeading - oldHeading;
			
			double oldPitch = headingAndPitch[1];
			double delPitch = -1 * (newPitch - oldPitch);

			win.updateObserverPerspectiveAnglesWithDeltas(delHeading, delPitch);
		}

	}

}

File: src/main/scala/prims/SetObserverAngles.scala
package view25d.prims

import org.nlogo.api.{ Argument, Command, Context }
import org.nlogo.core.Syntax

import view25d.View25DExtension

class SetObserverAngles extends Command {
  override def getSyntax: Syntax =
    Syntax.commandSyntax(right = List(Syntax.StringType, Syntax.NumberType, Syntax.NumberType),
                         agentClassString = "O---")

  override def perform(args: Array[Argument], context: Context): Unit = {
    val title = args(0).getString
    val heading = args(1).getDoubleValue
    val pitch = args(2).getDoubleValue

    View25DExtension.patchWindowMap.get(title).foreach(window => {
      val angles = window.getGLWindow.getObserverPerspectiveAngles

      window.getGLWindow.updateObserverPerspectiveAnglesWithDeltas(heading - angles(0), angles(1) - pitch)
    })

    View25DExtension.turtleWindowMap.get(title).foreach(window => {
      val angles = window.getGLWindow.getObserverPerspectiveAngles

      window.getGLWindow.updateObserverPerspectiveAnglesWithDeltas(heading - angles(0), angles(1) - pitch)
    })
  }
}
